[AutoDiff] Unify forward- and reverse-mode activity analysis.

dan-zheng opened the pull request:
- Unify handling of `apply` direct results.
  - Create `forEachApplyDirectResult` helper for iterating over `apply` direct
    results, with special logic for handling `destructure_tuple` users.
    This replaces `collectAllExtractedElements`, which required inefficient
    allocation.
  - Remove `collectAllActualResultsInTypeOrder` "indirect results" argument,
    which always comes from the `ApplyInst` argument.
- Unify forward- and reverse-mode activity analysis.
  - Remove `AutoDiffAssociatedFunctionKind` from `DifferentiableActivityInfo`.
- [TF-813](https://bugs.swift.org/projects/TF/browse/TF-813): verify that enabling forward-mode does not affect reverse-mode.
- Miscellaneous gardening.

Resolves [TF-800](https://bugs.swift.org/projects/TF/browse/TF-800) and [TF-813](https://bugs.swift.org/projects/TF/browse/TF-813).
Remaining differences between forward- and reverse-mode are expected.

---

Robust activity analysis simplification, deprecating https://github.com/apple/swift/pull/27356.

dan-zheng committed:
[AutoDiff] Unify forward- and reverse-mode activity analysis.

- Unify handling of `apply` direct results.
  - Use `collectAllExtractedElements` to collect `apply` direct results,
    with special logic for handling `destructure_tuple` users.
  - Remove `collectAllActualResultsInTypeOrder` "indirect results" argument,
    which always comes from the `ApplyInst` argument.
- Unify forward- and reverse-mode activity analysis.
  - Remove `AutoDiffAssociatedFunctionKind` from `DifferentiableActivityInfo`.
- TF-813: verify that enabling forward-mode does not affect reverse-mode.
- Miscellaneous gardening.

Resolves TF-800 and TF-813.
Remaining differences between forward- and reverse-mode are expected.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1420,24 +1444,20 @@ static bool isDifferentiationParameter(SILArgument *argument,
+/// For an `apply` instruction with active results, compute:
+/// - The results of the `apply` instruction, in type order.
+/// - The set of minimal parameter and result indices for differentiating the
+///   `apply` instruction.
```
Note: `collectMinimalIndicesForFunctionCall` now computes `SmallVectorImpl<SILValue> &results` instead of taking it as an operand. In practice, this does not change efficiency (results are computed in both cases) but simplifies call sites.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -120,6 +120,33 @@ static DestructureTupleInst *getSingleDestructureTupleUser(SILValue value) {
+///   collect all `tuple_extract` users in `results`. Only extracted elements
+///   are collected; non-extracted elements are initialized in `results` as
+///   `SILValue()`.
+static void collectAllExtractedElements(SILValue value,
```
This function should not be necessary. There are only two cases to handle: 1. `value` is a tuple, in which case you get its single `destructure_tuple` user and its result list (of type `SILInstructionResultArray`), and 2. `value` is not a tuple, in which case you use `value` directly. It's more efficient to just use `if (auto *dti = getSingleDestructureTupleUser(value)) { ... } else { ... }` at call sites because it won't create a new `SmallVector`.

dan-zheng replied:
`That makes sense, I'll change call sites to use `if (auto *dti = getSingleDestructureTupleUser(value)) { ... } else { ... }`.

dan-zheng replied:
Good point. After discussion, we ended up on a `forEachApplyDirectResult` helper, [done here](https://github.com/apple/swift/pull/27358/commits/503d554893e514859f583af5b2d390e604e0a1ad).

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -120,6 +120,33 @@ static DestructureTupleInst *getSingleDestructureTupleUser(SILValue value) {
+/// - If it is not tuple-typed, add the value to `results`.
+/// - If it is tuple-typed and has a single `destructure_tuple` user, collect
+///   all `destructure_tuple` results in `results`.
+/// - If it is tuple-typed and does not have a single `destructure_tuple` user,
```
The `tuple_extract` case is relevant for manual `tuple_extract` support, which I plan to add back to https://github.com/apple/swift/pull/26709.

dan-zheng replied:
Never mind. After in-person discussion:
* Supporting `tuple_extract` differential/pullback generation visitors is desirable, even though no known reproducers generate `tuple_extract` as input to Differentiation transform. [Discussed here.](https://github.com/apple/swift/pull/26709#discussion_r326915894)
* Thus, handwriting a `tuple_extract` test case is necessary. The test case I've been using does `tuple_extract` to the result of an `apply`, which is not good (SILGen never generates this). Thus, we should use a different test case (`tuple` then `tuple_extract`).

dan-zheng replied:
After discussion, we decided to remove `tuple_extract` visitors for now. They can be re-added later. [This Gist](https://gist.github.com/dan-zheng/1343673d2d4d20d403306283b42d522b) shows cases where SILGen produces `tuple_extract`. I tried to rewrite some of them as differentiation test cases but couldn't (some involve differentiation of `@inout` parameters).

dan-zheng committed:
Create more efficient `forEachApplyDirectResult` helper.

Previusly, `collectAllExtractedElements` required unnecessary SmallVector
allocation. Now, `forEachApplyDirectResult` avoids allocation and is
specialized to `ApplyInst`.

dan-zheng commented:
@swift-ci Please test tensorflow
